### DISCUSSION TITLE
Improving PIXI Graphics Performance When Rendering Huge Number of Objects.

### DISCUSSION THREAD

**Discussion by Amansaxena001** [pixijs/pixijs#10521] [Q&A]
Hi guys, 
I have an app that renders a huge number of Pixi graphics and these graphics are just like HTML `div` which can hold children, you can expand/collapse them, etc by this what I mean here is the graphics are dynamic in nature and we change the dimensions of them on the go depending upon what kind event was fired and also depending upon their `children` each graphics will be different in size.
I have `mainComponent` that is basically a `PIXI.Container` where all the elements are drawn.

The problem starts to come when I am adding borders on these graphics objects then the FPS heavily drops as now we are doing two draw calls because of which extra work needs to be done by GPU in order to render those.

I have seen a few examples using `RenderTexture` ( [medium](https://medium.com/@bigtimebuddy/rendering-fast-graphics-with-pixijs-6f547895c08c) ), and `Sprites` to improve the performance but I am not able to understand how would those work in my scenario. 

I need help here in determining the right approach for this use case

**Comment by Amansaxena001** [comment]
@bigtimebuddy Here is the detailed explanation of the problem that I was trying to explain to you [here](https://github.com/pixijs/pixijs/discussions/10249#discussioncomment-9364521). I saw your reply and thanks but to be very sure can you please go through the above description  so that Im sure you understood the problem correctly🙏🏼

**Comment by Amansaxena001** [comment]
@ivanpopelyshev can you pls help here? its urgent

**Comment by ivanpopelyshev** [comment]
I'm working on my own project and doing pixi only in spare time. If you have specific problem, please tal;k to @bigtimebuddy and post a bounty. Polygonal shapes can be defeinitely speed-up even without renderTexture, just sprites, but it requires writing shaders, thats brain time, and i couldnt even port graphics-smooth to v8 yet

In current situation with pixi dev priorities, i really cant help, please help by talking to guys

**Comment by RecursiveVoid** [comment]
@Amansaxena001 How does the shapes look like? Here is my tip: 

Replace simple rectangles and lines with Sprite(1x1 texture) where you can basically adjust the width and height for the desired size. and use particle container (not yet avaliable in v8 but u can downgrade to v7)

The border you mean stroke? Are those Webgl native lines or default triangulated lines, will of course come with a cost. (native lines are not yet in v8 i guess, but still you can downgrade to v7) 

How many active shapes do you have in your scene? Do you have a viewport where you can zoom in and out?

**Reply by Amansaxena001** [reply]
@RecursiveVoid This is how the shapes are like and these can expand or collapse. ( I am not going into detail how we do that ). Each of these shapes can have a `border: 1px solid` like what we do in HTML. In pixi we use `lineStyle` to draw a solid line around them. The `lineStyle` is available on `Graphics` object only so nothing special I am doing here.

<img width="518" alt="Screenshot 2024-05-12 at 8 40 15 PM" src="https://github.com/pixijs/pixijs/assets/49619101/1162f9a0-781e-47c0-ae9f-ab4e8c084cc4">

About the active shapes even though I ve enabled `culling` that does gives some performance boost but the shapes could easily go 5k-10k. And yes we have a `Viewport` which has a container where all the elements are put and rendered and you can zoom-in/out too.

**Reply by RecursiveVoid** [reply]
is this a mock-up or this how it should look with the background with a lot of diagonal lines?

**Reply by Amansaxena001** [reply]
@RecursiveVoid its just a mock to explain how my graphics look like currently with parent child relationship

**Reply by RecursiveVoid** [reply]
Did you try using Sprites and pushing them to ParticleContainer? because what you show me looks bunch of  Graphics rectangles. And those can be easily replaceable with Sprites

**Reply by RecursiveVoid** [reply]
Another question, on zoom level/scale change do you want to keep the stroke width constant? or re-scalable?

**Reply by Amansaxena001** [reply]
@RecursiveVoid we can use sprites and while using them I was not able to render the child components using renderTexture correctly instead when I used `app.renderer.generateTexture` that actually capture the snapshot of everything including children which made it work in the end.

The sprites easily offered 3-4x performance boost of the bat.

**Comment by Amansaxena001** [comment]
Closing this discussion as the query is now resolved. Thanks, @RecursiveVoid for actively participating.
